libavformat/rtcenc.c

```diff
@@ -0,0 +1,2226 @@
+    /* The private key for DTLS handshake. */
+    EVP_PKEY *dtls_pkey;
+    /* The SSL certificate used for fingerprint in SDP and DTLS handshake. */
+    X509 *dtls_cert;
+    /* The fingerprint of certificate, used in SDP offer. */
+    char *dtls_fingerprint;
+
+    /**
+     * This represents the material used to build the SRTP master key. It is
+     * generated by DTLS and has the following layout:
+     *          16B         16B         14B             14B
+     *      client_key | server_key | client_salt | server_salt
+     */
+    uint8_t dtls_srtp_material[DTLS_SRTP_MASTER_KEY_LEN * 2];
+
+    /* Whether the timer should be reset. */
+    int dtls_should_reset_timer;
+    /* Whether the DTLS is done at least for us. */
+    int dtls_done_for_us;
+    /* The number of packets retransmitted for DTLS. */
+    int dtls_arq_packets;
+
+    /* The UDP transport is used for delivering ICE, DTLS and SRTP packets. */
+    URLContext *udp_uc;
+
+    /* The maximum number of retries for DTLS transmission. */
+    int dtls_arq_max;
+    /* The step start timeout in ms for DTLS transmission. */
+    int dtls_arq_timeout;
+    /**
+     * The size of RTP packet, should generally be set to MTU.
+     * Note that pion requires a smaller value, for example, 1200.
+     */
+    int pkt_size;
+} DTLSContext;
+
+static int openssl_gen_private_key(DTLSContext *ctx)
+{
+    int ret = 0;
+#if OPENSSL_VERSION_NUMBER < 0x30000000L /* OpenSSL 3.0 */
+    EC_GROUP *ecgroup = NULL;
+    EC_KEY* dtls_eckey = NULL;
+#else
+    const char *curve = "prime256v1";
+#endif
+    void *s1 = ctx->log_avcl;
+
+    /* Should use the curves in ClientHello.supported_groups, for example:
+     *      Supported Group: x25519 (0x001d)
+     *      Supported Group: secp256r1 (0x0017)
+     *      Supported Group: secp384r1 (0x0018)
+     * Note that secp256r1 in openssl is called NID_X9_62_prime256v1 or prime256v1 in string,
+     * not NID_secp256k1 or secp256k1 in string
+     */
+#if OPENSSL_VERSION_NUMBER < 0x30000000L /* OpenSSL 3.0 */
+    ctx->dtls_pkey = EVP_PKEY_new();
+    dtls_eckey = EC_KEY_new();
+    ecgroup = EC_GROUP_new_by_curve_name(NID_X9_62_prime256v1);
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L // v1.1.x
+    /* For openssl 1.0, we must set the group parameters, so that cert is ok. */
+    EC_GROUP_set_asn1_flag(ecgroup, OPENSSL_EC_NAMED_CURVE);
+#endif
+
+    if (EC_KEY_set_group(dtls_eckey, ecgroup) != 1) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: EC_KEY_set_group failed\n");
+        ret = AVERROR(EINVAL);
+        goto end;
+    }
+
+    if (EC_KEY_generate_key(dtls_eckey) != 1) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: EC_KEY_generate_key failed\n");
+        ret = AVERROR(EINVAL);
+        goto end;
+    }
+
+    if (EVP_PKEY_set1_EC_KEY(ctx->dtls_pkey, dtls_eckey) != 1) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: EVP_PKEY_set1_EC_KEY failed\n");
+        ret = AVERROR(EINVAL);
+        goto end;
+    }
+#else
+    ctx->dtls_pkey = EVP_EC_gen(curve);
```

Comment:
Derek Buitenhuis:

```
> +    ctx->dtls_pkey = dtls_pkey = EVP_EC_gen(curve);
> +    if (!dtls_pkey) {
> +        av_log(s1, AV_LOG_ERROR, "DTLS: EVP_EC_gen curve=%s failed\n", curve);
```

For this and all ther logs: Something more descriptive than "function_name failed:"
should be used.